### PR TITLE
Migrate mutations/validations to new folder webhooks

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Make project more clearly.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/kueue/issues/311

#### Special notes for your reviewer:
Use `CustomDefaulter` and `CustomValidator` instead.

### CODE REVIEW

File: apis/kueue/webhooks/clusterqueue_webhook.go [hunk 1]
```diff
@@ -30,57 +31,70 @@ import (
...
-var _ webhook.Defaulter = &ClusterQueue{}
+var _ webhook.CustomDefaulter = &ClusterQueueWebhook{}
+
+// Default implements webhook.CustomDefaulter so a webhook will be registered for the type
+func (w *ClusterQueueWebhook) Default(ctx context.Context, obj runtime.Object) error {
+	cq := obj.(*kueue.ClusterQueue)
 
-// Default implements webhook.Defaulter so a webhook will be registered for the type
-func (cq *ClusterQueue) Default() {
 	clusterQueueLog.Info("defaulter", "clusterQueue", klog.KObj(cq))
```
Comment (alculquicondor) [inline apis/kueue/webhooks/clusterqueue_webhook.go:67]: Let's make this V(5), like the others. Also, the message should be a sentence.

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#remove-string-formatting-from-log-message

File: apis/kueue/webhooks/resourceflavor_webhook.go [hunk 2]
```diff
@@ -14,59 +14,72 @@ See the License for the specific language governing permissions and
...
-var _ webhook.Defaulter = &ResourceFlavor{}
+var _ webhook.CustomDefaulter = &ResourceFlavorWebhook{}
+
+// Default implements webhook.CustomDefaulter so a webhook will be registered for the type
+func (w *ResourceFlavorWebhook) Default(ctx context.Context, obj runtime.Object) error {
+	rf := obj.(*kueue.ResourceFlavor)
 
-// Default implements webhook.Defaulter so a webhook will be registered for the type
-func (rf *ResourceFlavor) Default() {
 	resourceFlavorLog.Info("defaulter", "resourceFlavor", klog.KObj(rf))
```
Comment (alculquicondor) [inline apis/kueue/webhooks/resourceflavor_webhook.go:54]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#remove-string-formatting-from-log-message

File: apis/kueue/webhooks/workload_webhook.go [hunk 3]
```diff
@@ -14,67 +14,75 @@ See the License for the specific language governing permissions and
...
-var _ webhook.Defaulter = &Workload{}
+var _ webhook.CustomDefaulter = &WorkloadWebhook{}
 
-// Default implements webhook.Defaulter so a webhook will be registered for the type
-func (r *Workload) Default() {
-	workloadlog.V(5).Info("defaulter", "workload", klog.KObj(r))
+// Default implements webhook.CustomDefaulter so a webhook will be registered for the type
+func (w *WorkloadWebhook) Default(ctx context.Context, obj runtime.Object) error {
+	wl := obj.(*kueue.Workload)
+	workloadlog.V(5).Info("defaulter", "workload", klog.KObj(wl))
```
Comment (alculquicondor) [inline apis/kueue/webhooks/workload_webhook.go:54]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md#remove-string-formatting-from-log-message

File: pkg/constants/constants.go [hunk 4]
```diff
@@ -35,4 +35,6 @@ const (
 	// that do not specify any priority class and there is no priority class
 	// marked as default.
 	DefaultPriority = 0
+
+	DefaultPodSetName = "main"
```
Comment (alculquicondor) [inline pkg/constants/constants.go:39]: If a constant is a default for an API, I think it's better for it to be in the apis package.

**Review by alculquicondor — COMMENTED** [review]
I think this is great, but maybe we can go a step further?

validations and defaults can be in separate packages, similar to `k8s.io/pkg/apis`. But if you have limited time, I'm fine merging this as-is.

File: apis/kueue/webhooks/clusterqueue_webhook.go (hunk 1 above)
Comment (kerthcet) [reply apis/kueue/webhooks/clusterqueue_webhook.go:67]: > Let's make this V(5), like the others. Also, the message should be a sentence.

Should we? V(5) is for tracing, maybe V2 or V3, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use

File: pkg/constants/constants.go (hunk 4 above)
Comment (kerthcet) [reply pkg/constants/constants.go:39]: Make sense.

**Comment by kerthcet** [comment]
> I think this is great, but maybe we can go a step further?
> 
> validations and defaults can be in separate packages, similar to `k8s.io/pkg/apis`. But if you have limited time, I'm fine merging this as-is.

I think this is not necessary, those packages under k8s.io used for exporting, not the same here. And we have to construct    `ClusterQueueWebhook` for both Defaulter and Validator. I'd like to leave it as it is.

File: pkg/constants/constants.go (hunk 4 above)
Comment (kerthcet) [reply pkg/constants/constants.go:39]: Make sense.

File: apis/kueue/v1alpha1/zz_generated.deepcopy.go [hunk 5]
```diff
@@ -24,7 +24,7 @@ package v1alpha1
 import (
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
+	runtime "k8s.io/apimachinery/pkg/runtime"
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply apis/kueue/v1alpha1/zz_generated.deepcopy.go:27]: It's auto generated by `make generate`.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/kueue/v1alpha1/zz_generated.deepcopy.go:27]: weird... did we upgrade any of the tools? :shrug:

File: apis/kueue/webhooks/clusterqueue_webhook.go [hunk 6]
```diff
@@ -30,57 +31,72 @@ import (
...
-// Default implements webhook.Defaulter so a webhook will be registered for the type
-func (cq *ClusterQueue) Default() {
-	clusterQueueLog.Info("defaulter", "clusterQueue", klog.KObj(cq))
-	if !controllerutil.ContainsFinalizer(cq, ResourceInUseFinalizerName) {
-		controllerutil.AddFinalizer(cq, ResourceInUseFinalizerName)
+// Default implements webhook.CustomDefaulter so a webhook will be registered for the type
+func (w *ClusterQueueWebhook) Default(ctx context.Context, obj runtime.Object) error {
+	cq := obj.(*kueue.ClusterQueue)
+
+	clusterQueueLog.V(5).Info("Setting defaulter", "clusterQueue", klog.KObj(cq))
```
Comment (alculquicondor) [inline apis/kueue/webhooks/clusterqueue_webhook.go:67]:
```suggestion
	clusterQueueLog.V(5).Info("Applying defaults", "clusterQueue", klog.KObj(cq))
```

File: test/integration/webhook/v1alpha1/workload_test.go [hunk 7]
```diff
@@ -50,13 +50,13 @@ var _ = ginkgo.Describe("Workload defaulting webhook", func() {
 			workload := testing.MakeWorkload(workloadName, ns.Name).Obj()
```
Comment (alculquicondor) [inline test/integration/webhook/v1alpha1/workload_test.go:50]: But `MakeWorkload` already sets some defaults (in particular, the podset name). Maybe we should use `kueue.Workload` directly in this test.

Same for other API objects.

File: apis/kueue/v1alpha1/zz_generated.deepcopy.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kerthcet) [reply apis/kueue/v1alpha1/zz_generated.deepcopy.go:27]: The latest update to Makefile is 3 months ago.

**Pull request merged**